### PR TITLE
rpc, gui: bumpfee signer support

### PR DESCRIPTION
**Pull request by Sjors**
The `bumpfee` RPC call and GUI fee bump interface now work with an external signer.

### CODE REVIEW

**Comment by Sjors** [comment]
I'm not tremendously excited by this approach of overloading `bumpfee`. But it also seems wrong to use `psbtbumpfee` here for the simple case of a single signer than has all the required keys. I could add yet another RPC method `signerbumpfee`?

cc @achow101 @instagibbs

**Review by meshcollider — COMMENTED** [review]
utACK 0c0d293d31e96c9119e4793df66b375ec9f8495e

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3521,16 +3521,39 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
     // If wallet private keys are enabled, return the new transaction id,
     // otherwise return the base64-encoded unsigned PSBT of the new transaction.
```
Comment (jonatack) [inline src/wallet/rpcwallet.cpp:3522]: Do the merged changes in #22021 need to be updated with this pull?

Comment (Sjors) [reply src/wallet/rpcwallet.cpp:3522]: I'll do a rebase and also address the nits above.

Comment (Sjors) [reply src/wallet/rpcwallet.cpp:3522]: For now these comments are correct. In a multisig setup `bumpfee` will throw ` "Transaction incomplete. Try psbtbumpfee instead."`. We can improve that later.

**Comment by Sjors** [comment]
Rebased, addressed comments and added a test.

**Review by meshcollider — COMMENTED** [review]
re-utACK dda02ce4f99cff37bbd41373df6ecb4541ed66f5

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -4174,10 +4195,10 @@ static RPCHelpMan send()
             PartiallySignedTransaction psbtx(rawTx);
 
             // First fill transaction with our data without signing,
-            // so external signers are not asked sign more than once.
+            // so external signers are not asked to sign more than once.
             bool complete;
-            pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, false, true);
-            const TransactionError err = pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, true, false);
+            pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, false /* sign */, true /* bip32derivs */);
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp:4200]: Usually we put the meaning-comment before the value.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3530,16 +3530,37 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
...
+            }
 
-        result.pushKV("txid", txid.GetHex());
+            complete = FinalizeAndExtractPSBT(psbtx, mtx);
+
+            if (!complete) {
+                throw JSONRPCError(RPC_WALLET_ERROR, "Transaction incomplete. Try psbtbumpfee instead.");
+            }
+            CTransactionRef tx(MakeTransactionRef(std::move(mtx)));
+            result.pushKV("txid", tx->GetHash().GetHex());
```
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp:3552]: Prefer to assign `txid` here and do a single shared `pushKV` later.

**Review by luke-jr — APPROVED** [review]
utACK

**Comment by luke-jr** [comment]
On further reflection, any reason not to put these details in `feebumper::SignTransaction` so it's more easily reused in the GUI?

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (Sjors) [reply src/wallet/rpcwallet.cpp:4200]: That indeed seems more common (`*/,` has 93 results vs 441 for `, /* `)

**Comment by Sjors** [comment]
@luke-jr good idea: now it works with GUI too!

**Comment by Sjors** [comment]
Rebased after #23667 (manually re-applied changes from `wallet/rpcwallet.cpp` to `rpc/spend.cpp`)

**Comment by achow101** [comment]
There was a bad rebase, you've re-added `src/wallet/rpcwallet.cpp`.

**Comment by Sjors** [comment]
Oops, fixed.

**Comment by achow101** [comment]
ACK d5f720fb48d3eeb64101af1d5614a3650643bb1e

**Comment by Sjors** [comment]
Rebased after #441, though not tested yet.

File: src/wallet/rpc/spend.cpp [hunk 4]
```diff
@@ -1121,10 +1124,10 @@ RPCHelpMan send()
             PartiallySignedTransaction psbtx(rawTx);
 
             // First fill transaction with our data without signing,
-            // so external signers are not asked sign more than once.
+            // so external signers are not asked to sign more than once.
             bool complete;
-            pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, false, true);
-            const TransactionError err = pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, true, false);
+            pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, /* sign */ false, /* bip32derivs */ true);
+            const TransactionError err = pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, /* sign */ true, /* bip32derivs */ false);
```
Comment (luke-jr) [inline src/wallet/rpc/spend.cpp:1130]:
```suggestion
            const TransactionError err = pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, /*sign=*/true, /*bip32derivs=*/false);
```

& other such changes

Comment (Sjors) [reply src/wallet/rpc/spend.cpp:1130]: Fixed (after the rebase moved everything around)

**Comment by Sjors** [comment]
Ok, it works again now.

File: src/wallet/feebumper.cpp [hunk 5]
```diff
@@ -239,7 +239,22 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
...
+        PartiallySignedTransaction psbtx(mtx);
+
+        // First fill transaction with our data without signing,
+        // so external signers are not asked to sign more than once.
+        bool complete;
+        wallet.FillPSBT(psbtx, complete, SIGHASH_ALL, false /* sign */, true /* bip32derivs */);
+        const TransactionError err = wallet.FillPSBT(psbtx, complete, SIGHASH_ALL, true /* sign */, false  /* bip32derivs */);
+        if (err != TransactionError::OK) return false;
+        complete = FinalizeAndExtractPSBT(psbtx, mtx);
+        return complete;
```
Comment (furszy) [inline src/wallet/feebumper.cpp:254]: nit: as this is the same code that we have in "rpc/send" --> `FinishTransaction`, could decouple it into a new function and use it from both sides.

**Review by furszy — COMMENTED** [review]
code review ACK 2c07cfac

**Review by jarolrod — COMMENTED** [review]
Approach ACK

I was able to successfully sign for a fee increase on my external signer. Want to continue to do more testing.

**Comment by achow101** [comment]
ACK 2c07cfacd1745844a1d3c57f2e8617549b9815d7

**Review by jarolrod — APPROVED** [review]
tACK 2c07cfa

I have tested and reviewed the code and agree it can be merged.

**Pull request merged**